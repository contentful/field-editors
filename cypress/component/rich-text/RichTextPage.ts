/* eslint-disable cypress/no-unnecessary-waiting */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import { INLINES } from '@contentful/rich-text-types';

import { cancelFakeDialog, confirmFakeDialog } from '../../fixtures';

export type EmbedType =
  | 'entry-block'
  | 'asset-block'
  | 'resource-block'
  | 'entry-inline'
  | 'resource-inline';

export class RichTextPage {
  get editor() {
    return cy.findByTestId('rich-text-editor').find('[data-slate-editor=true]');
  }

  get toolbar() {
    return {
      get headingsDropdown() {
        return cy.findByTestId('toolbar-heading-toggle');
      },

      toggleHeading(type: string) {
        this.headingsDropdown.click();
        cy.findByTestId(`dropdown-option-${type}`).click({ force: true });
      },

      get undo() {
        return cy.findByTestId('undo-toolbar-button');
      },

      get redo() {
        return cy.findByTestId('redo-toolbar-button');
      },

      get bold() {
        return cy.findByTestId('bold-toolbar-button');
      },

      get italic() {
        return cy.findByTestId('italic-toolbar-button');
      },

      get underline() {
        return cy.findByTestId('underline-toolbar-button');
      },

      get code() {
        return cy.findByTestId('code-toolbar-button');
      },

      get ul() {
        return cy.findByTestId('ul-toolbar-button');
      },

      get ol() {
        return cy.findByTestId('ol-toolbar-button');
      },

      get quote() {
        return cy.findByTestId('quote-toolbar-button');
      },

      get hr() {
        return cy.findByTestId('hr-toolbar-button');
      },

      get hyperlink() {
        return cy.findByTestId('hyperlink-toolbar-button');
      },

      get table() {
        return cy.findByTestId('table-toolbar-button');
      },

      get embedDropdown() {
        return cy.findByTestId('toolbar-entity-dropdown-toggle');
      },

      embed(type: EmbedType, autoConfirm: boolean = true) {
        this.embedDropdown.click();
        cy.findByTestId(`toolbar-toggle-embedded-${type}`).click();
        if (autoConfirm) {
          confirmFakeDialog();
        }
      },
    };
  }

  get forms() {
    return {
      get hyperlink() {
        return new HyperLinkModal();
      },
      get embed() {
        return {
          cancel: cancelFakeDialog,
          confirm: confirmFakeDialog,
        };
      },
    };
  }

  getValue() {
    cy.wait(500);

    return cy.getRichTextField().then((field) => {
      return field.getValue();
    });
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  expectValue(expectedValue: any) {
    // we want to make sure any kind of debounced behavior
    // is already triggered before we go on and assert the
    // content of the field in any test. Using cy.clock()
    // doesn't work for some reason
    // eslint-disable-next-line
    cy.wait(500);

    cy.getRichTextField().should((field) => {
      expect(field.getValue()).to.deep.equal(expectedValue);
    });
  }
}

class HyperLinkModal {
  get linkText() {
    return cy.findByTestId('link-text-input');
  }

  get linkType() {
    return cy.findByTestId('link-type-input');
  }

  setLinkType = (
    type:
      | INLINES.HYPERLINK
      | INLINES.ENTRY_HYPERLINK
      | INLINES.ASSET_HYPERLINK
      | INLINES.RESOURCE_HYPERLINK
  ) => {
    this.linkType.select(type);
  };

  get linkTarget() {
    return cy.findByTestId('link-target-input');
  }

  get linkEntityTarget() {
    return cy.findByTestId('entity-selection-link');
  }

  get submit() {
    return cy.findByTestId('confirm-cta');
  }

  get cancel() {
    return cy.findByTestId('cancel-cta');
  }
}
