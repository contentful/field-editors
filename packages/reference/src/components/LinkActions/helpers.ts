import { Asset, ContentType, EntityType, Entry, FieldExtensionSDK } from '../../types';
import { EditorPermissions } from '../../common/useEditorPermissions';

const getContentTypeIds = (contentTypes: ContentType[]) => contentTypes.map((ct) => ct.sys.id);

export async function createEntity(props: {
  sdk: FieldExtensionSDK;
  entityType: EntityType;
  contentTypeId?: string;
}) {
  if (props.entityType === 'Entry') {
    if (!props.contentTypeId) {
      return {};
    }
    const { entity, slide } = await props.sdk.navigator.openNewEntry<Entry>(props.contentTypeId, {
      slideIn: true,
    });
    return { entity, slide };
  } else {
    const { entity, slide } = await props.sdk.navigator.openNewAsset<Asset>({
      slideIn: true,
    });
    return { entity, slide };
  }
}

export async function selectSingleEntity(props: {
  sdk: FieldExtensionSDK;
  entityType: EntityType;
  editorPermissions: EditorPermissions;
}) {
  if (props.entityType === 'Entry') {
    return await props.sdk.dialogs.selectSingleEntry<Entry>({
      locale: props.sdk.field.locale,
      contentTypes: getContentTypeIds(props.editorPermissions.readableContentTypes),
    });
  } else {
    return props.sdk.dialogs.selectSingleAsset<Asset>({
      locale: props.sdk.field.locale,
      mimetypeGroups: props.editorPermissions.validations.mimetypeGroups,
    });
  }
}

export async function selectMultipleEntities(props: {
  sdk: FieldExtensionSDK;
  entityType: EntityType;
  editorPermissions: EditorPermissions;
}) {
  const value = props.sdk.field.getValue();

  const linkCount = Array.isArray(value) ? value.length : value ? 1 : 0;

  // TODO: Why not always set `min: 1` by default? Does it make sense to enforce
  //  user to select as many entities as the field's "min" requires? What if e.g.
  // "min" is 4 and the user wants to insert 2 entities first, then create 2 new ones?
  const min = Math.max(
    (props.editorPermissions.validations.numberOfLinks?.min || 1) - linkCount,
    1
  );
  // TODO: Consider same for max. If e.g. "max" is 4, we disable the button if the
  //  user wants to select 5 but we show no information why the button is disabled.
  const max = (props.editorPermissions.validations.numberOfLinks?.max || +Infinity) - linkCount;

  if (props.entityType === 'Entry') {
    return await props.sdk.dialogs.selectMultipleEntries<Entry>({
      locale: props.sdk.field.locale,
      contentTypes: getContentTypeIds(props.editorPermissions.readableContentTypes),
      min,
      max,
    });
  } else {
    return props.sdk.dialogs.selectMultipleAssets<Asset>({
      locale: props.sdk.field.locale,
      mimetypeGroups: props.editorPermissions.validations.mimetypeGroups,
      min,
      max,
    });
  }
}
