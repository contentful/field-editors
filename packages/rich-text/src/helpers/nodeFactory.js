export const document = (...content) => ({
  nodeType: 'document',
  content,
  data: {}
});

export const block = (nodeType, data, ...content) => ({
  nodeType,
  content,
  data
});

export const inline = (nodeType, data, ...content) => ({
  nodeType,
  content,
  data
});

export const text = (value = '', marks = []) => ({
  nodeType: 'text',
  value,
  marks,
  data: {}
});

export const mark = type => ({
  type
});
