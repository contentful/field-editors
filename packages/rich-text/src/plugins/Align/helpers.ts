import { ALIGNMENT, CustomElement } from './types';
import { HotkeyPlugin, isMarkActive, KeyboardHandler, toggleMark } from '@udecode/plate-core';
import { Node, Range } from 'slate';
import isHotkey from 'is-hotkey';

import { RichTextEditor } from '../../types';
import { Alignment } from '@udecode/plate-alignment';

  export const isAlignActive = (editor: RichTextEditor, type: Alignment): boolean => {
    const { selection } = editor;
  
    if (!selection) {
      return false;
    }

    // TODO: use Range to figure out if items use custom alignment
    if (Range.isExpanded(selection)) {
      const [start, end] = Range.edges(selection);
      const node = Node.common(editor, start.path, end.path);
      if ((node[0] as CustomElement).type === type) {
        return true
      }
    }

    return false



    // if (Range.isExpanded(selection)) {
    //   const [start, end] = Range.edges(selection);
    //   const node = Node.common(editor, start.path, end.path);
    //   console.log((node[0] as CustomElement))
    //   console.log('got to true')
  
    //   if ((node[0] as CustomElement).type === type) {
    //     return true;
    //   }
  };