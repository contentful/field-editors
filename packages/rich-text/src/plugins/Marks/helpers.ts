import { MARKS } from '@contentful/rich-text-types';
import isHotkey from 'is-hotkey';

import { isMarkActive } from '../../internal/queries';
import { toggleMark } from '../../internal/transforms';
import { PlateEditor, HotkeyPlugin, KeyboardHandler } from '../../internal/types';

export const toggleMarkAndDeactivateConflictingMarks = (editor: PlateEditor, mark: MARKS) => {
  const subs = [MARKS.SUPERSCRIPT, MARKS.SUBSCRIPT];
  const clear = subs.includes(mark) ? subs : [];
  toggleMark(editor, { key: mark, clear });
};

export const buildMarkEventHandler =
  (type: MARKS): KeyboardHandler<HotkeyPlugin> =>
  (editor, { options: { hotkey } }) =>
  (event) => {
    if (editor.selection && hotkey && isHotkey(hotkey, event)) {
      event.preventDefault();

      const isActive = isMarkActive(editor, type);
      editor.tracking.onShortcutAction(isActive ? 'unmark' : 'mark', { markType: type });
      toggleMarkAndDeactivateConflictingMarks(editor, type);
    }
  };
