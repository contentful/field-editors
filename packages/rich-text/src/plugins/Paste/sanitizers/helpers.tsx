import { SPEditor } from '@udecode/plate-core';

export type SanitizerTuple = [Document, SPEditor];

type Predicate = (node: ChildNode) => boolean;

const removeChildNodes = (node: ChildNode, predicate: Predicate) =>
  Array.from(node.childNodes)
    .filter(predicate)
    .forEach((table) => node.removeChild(table));

export const removeChildNodesUsingPredicate =
  (predicate: Predicate) =>
  (nodeList: NodeList): Node[] => {
    const nodes = Array.from(nodeList);
    while (nodes.length > 0) {
      const node = nodes.pop() as ChildNode;
      removeChildNodes(node, predicate);
      for (const childNode of Array.from(node.childNodes)) {
        nodes.push(childNode);
      }
    }
    return nodes;
  };
