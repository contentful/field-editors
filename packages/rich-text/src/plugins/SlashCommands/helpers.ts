export function getCaretTopPoint() {
  const sel = document.getSelection();
  if (!sel) return;
  const r = sel.getRangeAt(0);
  let rect;
  let r2;
  // supposed to be textNode in most cases
  // but div[contenteditable] when empty
  const node = r.startContainer;
  const offset = r.startOffset;
  if (offset > 0) {
    // new range, don't influence DOM state
    r2 = document.createRange();
    r2.setStart(node, offset - 1);
    r2.setEnd(node, offset);
    // https://developer.mozilla.org/en-US/docs/Web/API/range.getBoundingClientRect
    // IE9, Safari?(but look good in Safari 8)
    rect = r2.getBoundingClientRect();
    return { left: rect.right, top: rect.top };
    // @ts-expect-error
  } else if (offset < node.length) {
    r2 = document.createRange();
    // similar but select next on letter
    r2.setStart(node, offset);
    r2.setEnd(node, offset + 1);
    rect = r2.getBoundingClientRect();
    return { left: rect.left, top: rect.top };
  } else {
    // textNode has length
    // https://developer.mozilla.org/en-US/docs/Web/API/Element.getBoundingClientRect
    // @ts-expect-error
    rect = node.getBoundingClientRect();
    // @ts-expect-error
    const styles = getComputedStyle(node);
    const lineHeight = parseInt(styles.lineHeight);
    const fontSize = parseInt(styles.fontSize);
    // roughly half the whitespace... but not exactly
    const delta = (lineHeight - fontSize) / 2;
    return { left: rect.left, top: rect.top + delta };
  }
}

export function closePanel(editorId: string) {
  document.dispatchEvent(
    new CustomEvent('close-rte-palette-commands', {
      detail: {
        editorId,
      },
    })
  );
}

export function openPanel(editorId: string) {
  document.dispatchEvent(
    new CustomEvent('open-rte-palette-commands', {
      detail: {
        editorId,
      },
    })
  );
}
