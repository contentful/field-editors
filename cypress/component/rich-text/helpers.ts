import { BLOCKS } from '@contentful/rich-text-types';

import { block, text } from '../../../packages/rich-text/src/helpers/nodeFactory';

export const KEYS = {
  enter: { code: 'Enter', keyCode: 13, which: 13, key: 'Enter' },
  backspace: { code: 'Backspace', keyCode: 8, which: 8, key: 'Backspace' },
  tab: { code: 'Tab', keyCode: 9, which: 9, key: 'Tab' },
  delete: { code: 'Delete', keyCode: 8, which: 8, key: 'Delete' },
};

const buildHelper =
  (type) =>
  (...children) =>
    block(type, {}, ...children);
// Paragraphs
const paragraph = buildHelper(BLOCKS.PARAGRAPH);
export const paragraphWithText = (t) => paragraph(text(t, []));
export const emptyParagraph = () => paragraphWithText('');

// Tables
export const table = buildHelper(BLOCKS.TABLE);
export const row = buildHelper(BLOCKS.TABLE_ROW);
export const cell = buildHelper(BLOCKS.TABLE_CELL);
export const header = buildHelper(BLOCKS.TABLE_HEADER_CELL);
export const emptyCell = () => cell(emptyParagraph());
export const emptyHeader = () => header(emptyParagraph());
export const cellWithText = (t) => cell(paragraphWithText(t));
export const headerWithText = (t) => header(paragraphWithText(t));

// References
export const entryBlock = () =>
  block(BLOCKS.EMBEDDED_ENTRY, {
    target: {
      sys: {
        id: 'published-entry',
        type: 'Link',
        linkType: 'Entry',
      },
    },
  });

export const assetBlock = () =>
  block(BLOCKS.EMBEDDED_ASSET, {
    target: {
      sys: {
        id: 'published_asset',
        type: 'Link',
        linkType: 'Asset',
      },
    },
  });
