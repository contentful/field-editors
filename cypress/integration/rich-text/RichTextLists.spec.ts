/* eslint-disable mocha/no-setup-in-describe */
import { BLOCKS } from '@contentful/rich-text-types';

import { document as doc, block, text } from '../../../packages/rich-text/src/helpers/nodeFactory';
import { RichTextPage } from './RichTextPage';

describe('Rich Text Lists', () => {
  let richText: RichTextPage;

  const buildHelper =
    (type) =>
    (...children) =>
      block(type, {}, ...children);
  const paragraph = buildHelper(BLOCKS.PARAGRAPH);
  const paragraphWithText = (t) => paragraph(text(t, []));
  const emptyParagraph = () => paragraphWithText('');

  const entryBlock = () =>
    block(BLOCKS.EMBEDDED_ENTRY, {
      target: {
        sys: {
          id: 'example-entity-id',
          type: 'Link',
          linkType: 'Entry',
        },
      },
    });

  const assetBlock = () =>
    block(BLOCKS.EMBEDDED_ASSET, {
      target: {
        sys: {
          id: 'example-entity-id',
          type: 'Link',
          linkType: 'Asset',
        },
      },
    });

  const keys = {
    tab: { keyCode: 9, which: 9, key: 'Tab' },
  };

  function addBlockquote(content = '') {
    richText.editor.click().type(content);

    richText.toolbar.quote.click();

    const expectedValue = doc(
      block(BLOCKS.QUOTE, {}, block(BLOCKS.PARAGRAPH, {}, text(content, []))),
      block(BLOCKS.PARAGRAPH, {}, text('', []))
    );

    richText.expectValue(expectedValue);

    return expectedValue;
  }

  beforeEach(() => {
    richText = new RichTextPage();
    richText.visit();
  });

  const lists = [
    {
      getList: () => richText.toolbar.ul,
      listType: BLOCKS.UL_LIST,
      label: 'Unordered List (UL)',
    },
    { getList: () => richText.toolbar.ol, listType: BLOCKS.OL_LIST, label: 'Ordered List (OL)' },
  ];

  lists.forEach((test) => {
    describe(test.label, () => {
      it('should be visible', () => {
        test.getList().should('be.visible');
      });

      it('should add a new list', () => {
        richText.editor.click();

        test.getList().click();
        // TODO: Find a way to test deeper lists
        /*
					Having issues with `.type('{enter})` to break lines.
					The error is:
					Cannot resolve a Slate node from DOM node: [object HTMLSpanElement]
				*/
        richText.editor.type('item 1');

        const expectedValue = doc(
          block(
            test.listType,
            {},
            block(BLOCKS.LIST_ITEM, {}, block(BLOCKS.PARAGRAPH, {}, text('item 1', [])))
          ),
          emptyParagraph()
        );

        richText.expectValue(expectedValue);
      });

      it('should untoggle the list', () => {
        richText.editor.click();

        test.getList().click();

        richText.editor.type('some text');

        test.getList().click();

        const expectedValue = doc(
          block(BLOCKS.PARAGRAPH, {}, text('some text', [])),
          emptyParagraph()
        );

        richText.expectValue(expectedValue);
      });

      it('should unwrap blockquote', () => {
        addBlockquote('some text');

        test.getList().click();

        const expectedValue = doc(
          block(
            test.listType,
            {},
            block(BLOCKS.LIST_ITEM, {}, block(BLOCKS.PARAGRAPH, {}, text('some text', [])))
          ),
          emptyParagraph()
        );

        richText.expectValue(expectedValue);
      });

      it('should allow heading as direct child of <li>', () => {
        richText.editor.click();
        test.getList().click();

        richText.toolbar.toggleHeading(BLOCKS.HEADING_1);
        richText.editor.type('heading');

        const expectedValue = doc(
          block(
            test.listType,
            {},
            block(BLOCKS.LIST_ITEM, {}, block(BLOCKS.HEADING_1, {}, text('heading')))
          ),
          emptyParagraph()
        );

        richText.expectValue(expectedValue);
      });

      it('should allow HR as direct child of <li>', () => {
        richText.editor.click();
        test.getList().click();

        richText.toolbar.hr.click();

        const expectedValue = doc(
          block(test.listType, {}, block(BLOCKS.LIST_ITEM, {}, block(BLOCKS.HR, {}))),
          emptyParagraph()
        );

        richText.expectValue(expectedValue);
      });

      it('should allow embedded entry as direct child of <li>', () => {
        richText.editor.click();
        test.getList().click();

        richText.toolbar.embed('entry-block');

        const expectedValue = doc(
          block(test.listType, {}, block(BLOCKS.LIST_ITEM, {}, entryBlock())),
          emptyParagraph()
        );

        richText.expectValue(expectedValue);
      });

      it('should allow embedded asset as direct child of <li>', () => {
        richText.editor.click();
        test.getList().click();

        richText.toolbar.embed('asset-block');

        const expectedValue = doc(
          block(test.listType, {}, block(BLOCKS.LIST_ITEM, {}, assetBlock())),
          emptyParagraph()
        );

        richText.expectValue(expectedValue);
      });

      it('should allow block quotes as direct child of <li>', () => {
        richText.editor.click();
        test.getList().click();

        richText.toolbar.quote.click();
        richText.editor.type('quote');

        const expectedValue = doc(
          block(
            test.listType,
            {},
            block(BLOCKS.LIST_ITEM, {}, block(BLOCKS.QUOTE, {}, paragraphWithText('quote')))
          ),
          emptyParagraph()
        );

        richText.expectValue(expectedValue);
      });

      it('should preserve current marks when inserting a new li', () => {
        richText.editor.click();
        test.getList().click();

        // should only be applied to the word "bold "
        richText.toolbar.bold.click();
        richText.editor.type('bold ');
        richText.toolbar.bold.click();

        // should be applied to the rest of line AND next li
        richText.toolbar.italic.click();
        richText.editor.type('italic');

        richText.editor.type('{enter}');
        richText.editor.type('more italic text');

        richText.expectSnapshotValue();
      });

      it('should move nested list items when parent is invalid', () => {
        richText.editor.click();
        test.getList().click();

        richText.editor
          .type('1{enter}2{enter}3{enter}4')
          .trigger('keydown', keys.tab)
          .type('{uparrow}{uparrow}')
          .trigger('keydown', keys.tab)
          .type('{downarrow}{backspace}{backspace}');

        richText.expectSnapshotValue();
      });
    });
  });
});
