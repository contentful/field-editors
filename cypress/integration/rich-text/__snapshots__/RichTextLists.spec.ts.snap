exports[`Rich Text Lists > Unordered List (UL) > should preserve current marks when inserting a new li #0`] =
{
  "content": [
    {
      "content": [
        {
          "content": [
            {
              "content": [
                {
                  "data": {},
                  "marks": [
                    {
                      "type": "bold"
                    }
                  ],
                  "nodeType": "text",
                  "value": "bold "
                },
                {
                  "data": {},
                  "marks": [
                    {
                      "type": "italic"
                    }
                  ],
                  "nodeType": "text",
                  "value": "italic"
                }
              ],
              "data": {},
              "nodeType": "paragraph"
            }
          ],
          "data": {},
          "nodeType": "list-item"
        },
        {
          "content": [
            {
              "content": [
                {
                  "data": {},
                  "marks": [
                    {
                      "type": "italic"
                    }
                  ],
                  "nodeType": "text",
                  "value": "more italic text"
                }
              ],
              "data": {},
              "nodeType": "paragraph"
            }
          ],
          "data": {},
          "nodeType": "list-item"
        }
      ],
      "data": {},
      "nodeType": "unordered-list"
    },
    {
      "content": [
        {
          "data": {},
          "marks": [],
          "nodeType": "text",
          "value": ""
        }
      ],
      "data": {},
      "nodeType": "paragraph"
    }
  ],
  "data": {},
  "nodeType": "document"
};

exports[`Rich Text Lists > Ordered List (OL) > should preserve current marks when inserting a new li #0`] =
{
  "content": [
    {
      "content": [
        {
          "content": [
            {
              "content": [
                {
                  "data": {},
                  "marks": [
                    {
                      "type": "bold"
                    }
                  ],
                  "nodeType": "text",
                  "value": "bold "
                },
                {
                  "data": {},
                  "marks": [
                    {
                      "type": "italic"
                    }
                  ],
                  "nodeType": "text",
                  "value": "italic"
                }
              ],
              "data": {},
              "nodeType": "paragraph"
            }
          ],
          "data": {},
          "nodeType": "list-item"
        },
        {
          "content": [
            {
              "content": [
                {
                  "data": {},
                  "marks": [
                    {
                      "type": "italic"
                    }
                  ],
                  "nodeType": "text",
                  "value": "more italic text"
                }
              ],
              "data": {},
              "nodeType": "paragraph"
            }
          ],
          "data": {},
          "nodeType": "list-item"
        }
      ],
      "data": {},
      "nodeType": "ordered-list"
    },
    {
      "content": [
        {
          "data": {},
          "marks": [],
          "nodeType": "text",
          "value": ""
        }
      ],
      "data": {},
      "nodeType": "paragraph"
    }
  ],
  "data": {},
  "nodeType": "document"
};

exports[`Rich Text Lists > Unordered List (UL) > should move nested list items when parent is invalid #0`] =
{
  "content": [
    {
      "content": [
        {
          "content": [
            {
              "content": [
                {
                  "data": {},
                  "marks": [],
                  "nodeType": "text",
                  "value": "1"
                }
              ],
              "data": {},
              "nodeType": "paragraph"
            },
            {
              "content": [
                {
                  "content": [
                    {
                      "content": [
                        {
                          "data": {},
                          "marks": [],
                          "nodeType": "text",
                          "value": "2"
                        }
                      ],
                      "data": {},
                      "nodeType": "paragraph"
                    }
                  ],
                  "data": {},
                  "nodeType": "list-item"
                }
              ],
              "data": {},
              "nodeType": "unordered-list"
            }
          ],
          "data": {},
          "nodeType": "list-item"
        },
        {
          "content": [
            {
              "content": [
                {
                  "data": {},
                  "marks": [],
                  "nodeType": "text",
                  "value": "4"
                }
              ],
              "data": {},
              "nodeType": "paragraph"
            }
          ],
          "data": {},
          "nodeType": "list-item"
        }
      ],
      "data": {},
      "nodeType": "unordered-list"
    },
    {
      "content": [
        {
          "data": {},
          "marks": [],
          "nodeType": "text",
          "value": ""
        }
      ],
      "data": {},
      "nodeType": "paragraph"
    }
  ],
  "data": {},
  "nodeType": "document"
};

exports[`Rich Text Lists > Ordered List (OL) > should move nested list items when parent is invalid #0`] =
{
  "content": [
    {
      "content": [
        {
          "content": [
            {
              "content": [
                {
                  "data": {},
                  "marks": [],
                  "nodeType": "text",
                  "value": "1"
                }
              ],
              "data": {},
              "nodeType": "paragraph"
            },
            {
              "content": [
                {
                  "content": [
                    {
                      "content": [
                        {
                          "data": {},
                          "marks": [],
                          "nodeType": "text",
                          "value": "2"
                        }
                      ],
                      "data": {},
                      "nodeType": "paragraph"
                    }
                  ],
                  "data": {},
                  "nodeType": "list-item"
                }
              ],
              "data": {},
              "nodeType": "ordered-list"
            }
          ],
          "data": {},
          "nodeType": "list-item"
        },
        {
          "content": [
            {
              "content": [
                {
                  "data": {},
                  "marks": [],
                  "nodeType": "text",
                  "value": "4"
                }
              ],
              "data": {},
              "nodeType": "paragraph"
            }
          ],
          "data": {},
          "nodeType": "list-item"
        }
      ],
      "data": {},
      "nodeType": "ordered-list"
    },
    {
      "content": [
        {
          "data": {},
          "marks": [],
          "nodeType": "text",
          "value": ""
        }
      ],
      "data": {},
      "nodeType": "paragraph"
    }
  ],
  "data": {},
  "nodeType": "document"
};
